dns_server/src/crawler/core/mod.rs

```diff
@@ -0,0 +1,582 @@
+//! When the connection is closed, the DNS record is removed.
+//! When a connection fails, the server increases the backoff time between connection attempts.
+//! If the number of failed connection attempts exceeds the limit, the address is removed from the list.
+//! Once-reachable and newer-reachable addresses have different connection failure limits
+//! (equivalent to about 1 month and about 1 hour, respectively).
+//!
+//! The core is deterministic (and therefore can't do IO operations itself).
+//! All input comes in the form of `CoreEvent` messages and all output comes in the form of `CoreCommand` messages.
+
+use std::{
+    collections::{btree_map::Entry, BTreeMap, BTreeSet},
+    marker::PhantomData,
+    net::{IpAddr, Ipv4Addr, Ipv6Addr},
+    sync::Arc,
+    time::Duration,
+};
+
+use common::chain::ChainConfig;
+use crypto::random::{seq::IteratorRandom, Rng};
+use logging::log;
+use p2p::{
+    error::P2pError,
+    message::{AnnounceAddrRequest, PeerManagerMessage, PingRequest, PingResponse},
+    net::{
+        default_backend::transport::TransportAddress,
+        types::{ConnectivityEvent, PeerInfo},
+        NetworkingService,
+    },
+    peer_manager::global_ip::IsGlobalIp,
+    types::peer_address::PeerAddress,
+};
+
+/// How many outbound connection attempts can be made per heartbeat
+const MAX_CONNECTS_PER_HEARTBEAT: usize = 25;
+
+/// When the server drops the unreachable node address. Used for negative caching.
+const PURGE_UNREACHABLE_TIME: Duration = Duration::from_secs(3600);
+
+/// When the server drops the unreachable node address that was once reachable. This should take about a month.
+/// Such a long time is useful if the server itself has prolonged connectivity problems.
+pub const PURGE_REACHABLE_FAIL_COUNT: u32 = 35;
+
+/// Address state transition
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum AddressTransition {
+    Disconnected,
+    Connecting,
+    Connected,
+}
+
+/// Connection state of a potential node address (outbound only)
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum AddressState {
+    Disconnected {
+        /// The number of consecutive failed connection attempts.
+        /// New connection attempts are made after a progressive backoff time.
+        fail_count: u32,
+    },
+    Connecting {
+        /// Same as in the Disconnected state
+        fail_count: u32,
+    },
+    Connected {},
+}
+
+/// Additional state of a potential node address
+struct AddressData {
+    /// Connection state
+    state: AddressState,
+
+    /// Last time when the connection state was updated
+    state_updated_at: Duration,
+
+    /// Whether the address was reachable at least once.
+    /// Addresses that were once reachable are stored in the DB.
+    was_reachable: bool,
+
+    /// Whether the address was added from the command line
+    user_added: bool,
+}
+
+#[derive(Clone)]
+pub struct CrawlerCoreConfig {
+    pub p2p_port: u16,
```

Comment:
Aren't we not supposed to open a port for DNS in p2p?

Reply:
The server does not open any listening ports. This is the default p2p port on which nodes normally accept inbound connections. The name was a bit confusing. I have renamed it and added a comment.